In [1]:
import pandas as pd

In [2]:
games = pd.read_csv("nba_games.csv", index_col=0)

In [3]:
games.head()

mp  mp.1  fg  fga    fg%  3p  3pa    3p%  ft  fta  ...  tov%_max_opp  \
0  240   240  39   81  0.481   6   20  0.300  14   18  ...          22.8   
1  240   240  36  100  0.360   7   31  0.226  16   19  ...          50.0   
2  240   240  37   85  0.435   8   19  0.421  17   23  ...          20.0   
3  240   240  41   89  0.461   8   21  0.381  17   19  ...          28.6   
4  240   240  27   86  0.314   6   26  0.231  15   20  ...          16.8   

   usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  home_opp  \
0          29.0           178           111       DAL         95         1   
1          32.6           152           111       ATL         98         0   
2          30.9           148           116       SAS        107         1   
3          30.9           138           118       MIN         99         0   
4          30.9           157            90       MEM         92         1   

   season        date    won  
0    2016  09/12/2015   True  
1    2016  09/12/2015  False  
2    2018  18/10/2017  False  
3    2018  18/10/2017   True  
4    2021  30/04/2021  False  

[5 rows x 150 columns]

In [4]:
games.shape

(17772, 150)

In [5]:
games["team"].value_counts()

GSW    655
BOS    653
TOR    635
MIL    628
MIA    617
CLE    610
HOU    610
LAC    607
PHI    605
UTA    605
DEN    603
POR    603
OKC    595
DAL    593
SAS    593
ATL    590
PHO    590
MEM    585
IND    583
LAL    581
BRK    580
WAS    580
NOP    571
ORL    565
CHI    558
MIN    558
CHO    556
DET    556
SAC    554
NYK    553
Name: team, dtype: int64

In [6]:
games[games["team"] == "GSW"].sort_values("date")

mp  mp.1  fg  fga    fg%  3p  3pa    3p%  ft  fta  ...  tov%_max_opp  \
15567  240   240  31   79  0.392   7   35  0.200  29   33  ...          50.0   
16686  240   240  46   86  0.535  17   32  0.531  14   21  ...          17.6   
1633   240   240  47   96  0.490  21   48  0.438  11   14  ...          29.4   
1196   240   240  48   90  0.533  18   35  0.514  17   19  ...          27.4   
13810  240   240  43   96  0.448  20   46  0.435  18   18  ...          23.1   
...    ...   ...  ..  ...    ...  ..  ...    ...  ..  ...  ...           ...   
2358   240   240  46   84  0.548  17   30  0.567  25   35  ...          33.3   
12751  240   240  47   89  0.528  16   32  0.500  21   30  ...          31.6   
10092  240   240  43   84  0.512  11   23  0.478  17   27  ...          20.5   
8348   240   240  46   84  0.548  14   28  0.500  26   32  ...          50.0   
14560  265   265  44   96  0.458  12   27  0.444  13   18  ...          50.0   

       usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  \
15567          29.0           250           102       POR        123   
16686          29.2           163           127       UTA        116   
1633           30.8           231           131       CHO        111   
1196           28.7           139           139       CLE        112   
13810          34.6           230           138       SAS        120   
...             ...           ...           ...       ...        ...   
2358           36.1           218           131       NOP        120   
12751          26.8           149           132       NOP        121   
10092          34.3           157           124       HOU        110   
8348           44.5           137           134       PHO        109   
14560          25.9           175           120       SAS        117   

       home_opp  season        date    won  
15567         0    2021  01/01/2021  False  
16686         1    2022  01/01/2022   True  
1633          0    2017  01/02/2017   True  
1196          1    2020  01/02/2020   True  
13810         1    2022  01/02/2022   True  
...         ...     ...         ...    ...  
2358          1    2016  31/10/2015   True  
12751         0    2019  31/10/2018   True  
10092         1    2016  31/12/2015   True  
8348          1    2019  31/12/2018   True  
14560         1    2020  31/12/2019  False  

[655 rows x 150 columns]

Counts how many games are in each season 

In [7]:
games["season"].value_counts()

2022    2646
2016    2632
2018    2624
2019    2624
2017    2618
2021    2342
2020    2286
Name: season, dtype: int64

In [8]:
games.dtypes

mp             int64
mp.1           int64
fg             int64
fga            int64
fg%          float64
              ...   
total_opp      int64
home_opp       int64
season         int64
date          object
won             bool
Length: 150, dtype: object

Deleting column that is not needed 

In [9]:
del games["mp.1"]

Converting existing column to a date time then overiding the existing column with the date time.

In [10]:
games["date"] = pd.to_datetime(games["date"], format="%d/%m/%Y")

In [11]:
games

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  tov%_max_opp  \
0      240  39   81  0.481   6   20  0.300  14   18  0.778  ...          22.8   
1      240  36  100  0.360   7   31  0.226  16   19  0.842  ...          50.0   
2      240  37   85  0.435   8   19  0.421  17   23  0.739  ...          20.0   
3      240  41   89  0.461   8   21  0.381  17   19  0.895  ...          28.6   
4      240  27   86  0.314   6   26  0.231  15   20  0.750  ...          16.8   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  35   81  0.432  11   26  0.423  27   36  0.750  ...          34.2   
17768  240  37   74  0.500  13   25  0.520  26   37  0.703  ...          25.0   
17769  240  42   89  0.472  14   33  0.424  10   20  0.500  ...          25.6   
17770  240  41   85  0.482   9   26  0.346  26   30  0.867  ...          27.7   
17771  240  33   85  0.388  12   44  0.273  28   34  0.824  ...          51.5   

       usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  \
0              29.0           178           111       DAL         95   
1              32.6           152           111       ATL         98   
2              30.9           148           116       SAS        107   
3              30.9           138           118       MIN         99   
4              30.9           157            90       MEM         92   
...             ...           ...           ...       ...        ...   
17767          33.7           160           118       OKC         92   
17768          30.0           139           129       ORL        108   
17769          29.9           175           126       LAC        113   
17770          27.1           150           126       MIA        106   
17771          36.2           141           114       BOS        117   

       home_opp  season       date    won  
0             1    2016 2015-12-09   True  
1             0    2016 2015-12-09  False  
2             1    2018 2017-10-18  False  
3             0    2018 2017-10-18   True  
4             1    2021 2021-04-30  False  
...         ...     ...        ...    ...  
17767         0    2019 2018-10-19   True  
17768         1    2017 2016-12-14   True  
17769         0    2017 2016-12-14  False  
17770         1    2020 2020-09-19   True  
17771         0    2020 2020-09-19  False  

[17772 rows x 149 columns]

In [12]:
games.dtypes

mp                    int64
fg                    int64
fga                   int64
fg%                 float64
3p                    int64
                  ...      
total_opp             int64
home_opp              int64
season                int64
date         datetime64[ns]
won                    bool
Length: 149, dtype: object

Creating predictors for machine learning 
home says wheteher the team played at a home game or an away game

concatenates the home category in the table 

In [13]:
games ["venue_code"] = games["home"].astype("category").cat.codes

This creates the code for each opponent

In [14]:
games ["opp_code"] = games["team_opp"].astype("category").cat.codes

In [15]:
games ["day_code"] = games ["date"].dt.dayofweek

In [16]:
games

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  drtg_max_opp  \
0      240  39   81  0.481   6   20  0.300  14   18  0.778  ...           111   
1      240  36  100  0.360   7   31  0.226  16   19  0.842  ...           111   
2      240  37   85  0.435   8   19  0.421  17   23  0.739  ...           116   
3      240  41   89  0.461   8   21  0.381  17   19  0.895  ...           118   
4      240  27   86  0.314   6   26  0.231  15   20  0.750  ...            90   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  35   81  0.432  11   26  0.423  27   36  0.750  ...           118   
17768  240  37   74  0.500  13   25  0.520  26   37  0.703  ...           129   
17769  240  42   89  0.472  14   33  0.424  10   20  0.500  ...           126   
17770  240  41   85  0.482   9   26  0.346  26   30  0.867  ...           126   
17771  240  33   85  0.388  12   44  0.273  28   34  0.824  ...           114   

       team_opp  total_opp  home_opp  season       date    won  venue_code  \
0           DAL         95         1    2016 2015-12-09   True           0   
1           ATL         98         0    2016 2015-12-09  False           1   
2           SAS        107         1    2018 2017-10-18  False           0   
3           MIN         99         0    2018 2017-10-18   True           1   
4           MEM         92         1    2021 2021-04-30  False           0   
...         ...        ...       ...     ...        ...    ...         ...   
17767       OKC         92         0    2019 2018-10-19   True           1   
17768       ORL        108         1    2017 2016-12-14   True           0   
17769       LAC        113         0    2017 2016-12-14  False           1   
17770       MIA        106         1    2020 2020-09-19   True           0   
17771       BOS        117         0    2020 2020-09-19  False           1   

       opp_code  day_code  
0             6         2  
1             0         2  
2            26         2  
3            17         2  
4            14         4  
...         ...       ...  
17767        20         4  
17768        21         2  
17769        12         2  
17770        15         5  
17771         1         5  

[17772 rows x 152 columns]

The following code sets up a target to predict

In [26]:
games["target"] = (games["won"] == True).astype("int")

In [27]:
games

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  team_opp  \
0      240  39   81  0.481   6   20  0.300  14   18  0.778  ...       DAL   
1      240  36  100  0.360   7   31  0.226  16   19  0.842  ...       ATL   
2      240  37   85  0.435   8   19  0.421  17   23  0.739  ...       SAS   
3      240  41   89  0.461   8   21  0.381  17   19  0.895  ...       MIN   
4      240  27   86  0.314   6   26  0.231  15   20  0.750  ...       MEM   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
17767  240  35   81  0.432  11   26  0.423  27   36  0.750  ...       OKC   
17768  240  37   74  0.500  13   25  0.520  26   37  0.703  ...       ORL   
17769  240  42   89  0.472  14   33  0.424  10   20  0.500  ...       LAC   
17770  240  41   85  0.482   9   26  0.346  26   30  0.867  ...       MIA   
17771  240  33   85  0.388  12   44  0.273  28   34  0.824  ...       BOS   

       total_opp  home_opp  season       date    won  venue_code  opp_code  \
0             95         1    2016 2015-12-09   True           0         6   
1             98         0    2016 2015-12-09  False           1         0   
2            107         1    2018 2017-10-18  False           0        26   
3             99         0    2018 2017-10-18   True           1        17   
4             92         1    2021 2021-04-30  False           0        14   
...          ...       ...     ...        ...    ...         ...       ...   
17767         92         0    2019 2018-10-19   True           1        20   
17768        108         1    2017 2016-12-14   True           0        21   
17769        113         0    2017 2016-12-14  False           1        12   
17770        106         1    2020 2020-09-19   True           0        15   
17771        117         0    2020 2020-09-19  False           1         1   

       day_code  target  
0             2       1  
1             2       0  
2             2       0  
3             2       1  
4             4       0  
...         ...     ...  
17767         4       1  
17768         2       1  
17769         2       0  
17770         5       1  
17771         5       0  

[17772 rows x 153 columns]

Scaling data having vector X as [0], or standardise to have mean 0 and variance 1(ask about this)
StandardScaler is used for standardisation

In [19]:

from sklearn.preprocessing import StandardScaler

In [20]:
#Fitting 
#scaler =  StandardScaler()
#fitting inly in training data
#scaler.fit(games)
#games_train =  scaler.transform(games)
#apply same transformation to test data
#games_test = scaler.transform(games)

Multi-layer Perceptron (MLP) is a supervised learning algorithm that learns a function 
 by training on a dataset, where 
 is the number of dimensions for input and 
 is the number of dimensions for output. Given a set of features 
 and a target 
, it can learn a non-linear function approximator for either classification or regression. It is different from logistic regression, in that between the input and the output layer, there can be one or more non-linear layers, called hidden layers. Figure 1 shows a one hidden layer MLP with scalar output.
Reference: https://scikit-learn.org/stable/modules/neural_networks_supervised.html#tips-on-practical-use

This Class MLPClassifier implements a multi-layer perceptron (MLP) algorithm that trains using backpropagation(taking the error rate of a forward propagation and feeding this loss backwards through the neural network layer to fine-tune the weights, key to neural net training)
Reference: https://builtin.com/machine-learning/backpropagation-neural-network

In [21]:
from sklearn.neural_network import MLPClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

MLP trains on two arrays: array X of size (n_samples, n_features), which holds the training samples represented as floating point feature vectors; and array y of size (n_samples,), which holds the target values (class labels) for the training samples:

In [22]:
clf = MLPClassifier(random_state=1, max_iter=300)

In [23]:
#X, y = make_classification(n_samples=100, random_state=1)

In [24]:
#X is the column that we are predict from 

In [29]:
predictors = ["venue_code","opp_code","day_code"]
X_train = games[games["date"]<'2022-01-01']#["predictors"]#.astype("int")
y_train = games[games["date"]<'2022-01-01']#["target"]#.astype("int")
X_test = games[games["date"]>'2022-01-01']#["predictors"]#.astype("int")
y_test = games[games["date"]>'2022-01-01']#["target"]#.astype("int")

In [ ]:
 #X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state=1)

In [30]:
clf.fit(X_train[predictors], y_train["target"])
#X_train[targets], y_train[targets]

MLPClassifier(max_iter=300, random_state=1)

In [31]:
preds = clf.predict(X_test[predictors])
#X_test[predictors], y_test[predictors]

In [33]:
from sklearn.metrics import accuracy_score

In [34]:
acc = accuracy_score(X_test["target"], preds)
#X_test["target"], y_test["target"]

In [35]:
acc

0.5512658227848102

In [37]:
combined = pd.DataFrame(dict(actual=X_test["target"], prediction=preds))

In [38]:
pd.crosstab(index= combined["actual"],columns=combined["prediction"])

prediction    0    1
actual              
0           470  320
1           389  401

470 times the model was correct, 320 times the model was wrong 

In [39]:
from sklearn.metrics import precision_score

when a win was predicted the team only won 55% of the time which is low 

In [40]:
precision_score(X_test["target"], preds)

0.5561719833564494

Improvement of the model 
Improving the precison with rolling averages- by splitting the data set up in teams
This groups the data set into team statistics 

In [41]:
grouped_games = games.groupby("team")

In [42]:
group = grouped_games.get_group("LAL")

In [43]:
group

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  team_opp  \
37     240  34   82  0.415  11   40  0.275  17   23  0.739  ...       ORL   
44     240  39   78  0.500  10   30  0.333  28   43  0.651  ...       POR   
55     240  32   90  0.356   4   19  0.211  13   17  0.765  ...       LAC   
58     240  43   91  0.473   8   26  0.308  16   23  0.696  ...       SAS   
74     240  43   90  0.478  10   30  0.333  14   17  0.824  ...       DAL   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
17547  240  31   86  0.360   6   23  0.261  22   26  0.846  ...       PHO   
17585  240  43   86  0.500  14   29  0.483  23   28  0.821  ...       NOP   
17586  240  36   82  0.439   7   23  0.304  17   23  0.739  ...       POR   
17592  240  31   83  0.373   3   21  0.143  12   16  0.750  ...       GSW   
17615  240  47   85  0.553  19   41  0.463  26   30  0.867  ...       POR   

       total_opp  home_opp  season       date    won  venue_code  opp_code  \
37            93         0    2021 2021-03-28   True           1        21   
44           108         1    2020 2020-08-22   True           0        24   
55            91         0    2016 2016-04-06  False           1        12   
58           117         1    2022 2022-03-07  False           0        26   
74           115         1    2021 2021-04-22  False           0         6   
...          ...       ...     ...        ...    ...         ...       ...   
17547         95         0    2016 2016-03-18  False           1        23   
17585        113         0    2020 2020-01-03   True           1        18   
17586        108         1    2016 2015-11-28  False           0        24   
17592        111         1    2016 2015-11-24  False           0         9   
17615        106         0    2022 2021-12-31   True           1        24   

       day_code  target  
37            6       1  
44            5       1  
55            2       0  
58            0       0  
74            3       0  
...         ...     ...  
17547         4       0  
17585         4       1  
17586         5       0  
17592         1       0  
17615         4       1  

[581 rows x 153 columns]

The games of LAL (Los Angeles Lakers)if the prediction is on the fourth game the model should look at the previous three games to help make better predictions. For example if the team is on a losing streak or a winning streak